File: src/lens/gime_mapping_loader.py
import json
import hashlib
import os

ACTIVE_MAPPING = None

def load_gime_v0_1():
    global ACTIVE_MAPPING
    if ACTIVE_MAPPING is not None:
        return ACTIVE_MAPPING

    current_path = os.path.dirname(os.path.abspath(__file__))
    while not os.path.exists(os.path.join(current_path, 'assets')):
        parent = os.path.dirname(current_path)
        if parent == current_path:
            break
        current_path = parent
        
    filepath = os.path.join(current_path, 'assets', 'lenses', 'gime', 'v0.1', 'mapping.json')
    with open(filepath, 'r', encoding='utf-8') as f:
        raw = f.read()
    
    mapping = json.loads(raw)

    if mapping.get('lens') != 'GIME':
        raise Exception('Invalid lens')
    if mapping.get('version') != 'v0.1':
        raise Exception('Invalid version')

    stored_checksum = mapping.get('checksum')
    import re
    blanked = re.sub(r'"checksum":\s*".*?"', '"checksum": ""', raw)
    digest = hashlib.sha256(blanked.encode('utf-8')).hexdigest()

    if digest != stored_checksum:
        raise Exception('Checksum mismatch')
    
    import datetime
    boot_log = {
        "event": "lens_loaded",
        "lens_name": "GIME",
        "lens_version": "v0.1",
        "lens_checksum": stored_checksum,
        "timestamp": datetime.datetime.utcnow().isoformat() + "Z"
    }
    print(json.dumps(boot_log))

    ACTIVE_MAPPING = mapping
    return ACTIVE_MAPPING

if __name__ == "__main__":
    load_gime_v0_1()

File: services/l2-ingestion/src/lens/gime_mapping_loader.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

let activeMapping: any = null;

export const load_gime_v0_1 = (): any => {
    if (activeMapping) {
        return activeMapping;
    }

    let currentPath = __dirname;
    while (!fs.existsSync(path.join(currentPath, 'assets')) && currentPath !== path.parse(currentPath).root) {
        currentPath = path.dirname(currentPath);
    }
    const filePath = path.join(currentPath, 'assets', 'lenses', 'gime', 'v0.1', 'mapping.json');
    if (!fs.existsSync(filePath)) {
        throw new Error(`Mapping file not found at ${filePath}`);
    }

    const raw = fs.readFileSync(filePath, 'utf8');
    const mapping = JSON.parse(raw);

    // Validate header fields
    if (mapping.lens !== 'GIME') {
        throw new Error('Invalid lens name: expected GIME');
    }
    if (mapping.version !== 'v0.1') {
        throw new Error('Invalid lens version: expected v0.1');
    }

    // Checksum validation reproducing S9-W1-01 exact method
    const storedChecksum = mapping.checksum;
    const blanked = raw.replace(/"checksum": ".*"/, '"checksum": ""');

    // Checksum must match
    const computedHash = crypto.createHash('sha256').update(blanked).digest('hex');
    if (computedHash !== storedChecksum) {
        throw new Error(`Checksum mismatch. Expected: ${storedChecksum}, but got: ${computedHash}`);
    }

    // Cache the mapping in memory
    activeMapping = mapping;

    // Emit structured boot telemetry
    console.log(JSON.stringify({
        event: 'lens_loaded',
        lens_name: mapping.lens,
        lens_version: mapping.version,
        lens_checksum: storedChecksum,
        timestamp: new Date().toISOString()
    }));

    return activeMapping;
};

export const getActiveMapping = (): any => {
    if (!activeMapping) {
        throw new Error("Mapping not loaded yet. Boot sequence must complete first.");
    }
    return activeMapping;
};
